runtime/visualize.cc

```diff
@@ -0,0 +1,559 @@
+    "    flat_min += (extent - 1) * min(0, dims[i].stride);\n"
+    "    flat_max += (extent - 1) * max(0, dims[i].stride);\n"
+    "  }\n"
+    "  let size = flat_max - flat_min + elem_size;\n"
+    "  let base = flat_allocate(size);\n"
+    "  define_mapping(name, base, size, elem_size, dims);\n"
+    "  return {base: base, elem_size: elem_size, dims: dims, color: next_color()};\n"
+    "}\n"
+    "function free(b) {}\n"
+    "function make_buffer(base, elem_size, dims) { return {base: base, elem_size: elem_size, dims: dims, color: "
+    "next_color()}; }\n"
+    "function clone_buffer(b) { return structuredClone(b); }\n"
+    "function flat_offset_dim(d, x) { return ((x - d.bounds.min) % d.fold_factor) * d.stride; }\n"
+    "function crop_dim(b, d, bounds) {\n"
+    "  let result = clone_buffer(b);\n"
+    "  let new_min = max(b.dims[d].bounds.min, bounds.min);\n"
+    "  let new_max = min(b.dims[d].bounds.max, bounds.max);\n"
+    "  b.base += flat_offset_dim(b.dims[d], new_min);\n"
+    "  b.dims[d].bounds.min = new_min;\n"
+    "  b.dims[d].bounds.max = new_max;\n"
+    "  return result;\n"
+    "}\n"
+    "function crop_buffer(b, bounds) {\n"
+    "  let result = clone_buffer(b);\n"
+    "  for (let d = 0; d < bounds.length; ++d) {\n"
+    "    crop_dim(b, d, bounds[d]);\n"
+    "  }\n"
+    "  return result;\n"
+    "}\n"
+    "function slice_dim(b, d, at) {\n"
+    "  let result = clone_buffer(b);\n"
+    "  b.base += flat_offset_dim(b.dims[d], at);\n"
+    "  b.dims.splice(d, 1);\n"
+    "  return result;\n"
+    "}\n"
+    "function slice_buffer(b, at) {\n"
+    "  let result = clone_buffer(b);\n"
+    "  for (let d = at.length - 1; d >= 0; --d) {\n"
+    "    slice_dim(b, d, at[d]);\n"
+    "  }\n"
+    "  return result;\n"
+    "}\n"
+    "function truncate_rank(b, rank) {\n"
+    "  let result = clone_buffer(b);\n"
+    "  b.dims.length = rank;\n"
+    "  return result;\n"
+    "}\n"
+    "function produce(b) {\n"
+    "  for (let m of __heap_map) {\n"
+    "    if (m.begin <= b.base && b.base < m.end) {\n"
+    "      m.element.productions.push({t: __event_t++, buf: clone_buffer(b)});\n"
+    "      return;\n"
+    "    }\n"
+    "  }\n"
+    "}\n"
+    "function consume(b) {}\n";
+const char* footer =
+    "let __end_t = __event_t;\n"
+    "let __event_slider = document.getElementById('event_slider');\n"
+    "let __autoplay = true;\n"
+    "__event_slider.max = __end_t - 1;\n"
+    "document.addEventListener('keyup', event => { if (event.code === 'Space') __autoplay = !__autoplay; });\n"
+    "setInterval(function() {\n"
+    "  if (__autoplay) {\n"
+    "    __current_t = (__current_t + 1) % __end_t;\n"
+    "    __event_slider.value = __current_t;\n"
+    "  }\n"
+    "}, 100);\n"
+    "</script></body></html>\n";
+
+}  // namespace
+
+void visualize(const char* filename, const pipeline& p, pipeline::scalars args, pipeline::buffers inputs,
```

Comment:
general q: it's unclear to me at this point in my limited knowledge of the codebase why some of our classes are passed by const ref (eg pipeline) and some are passed by copy (eg pipeline::scalars), not sure if comments are necessary here or if all we be clear upon further knowledge but it's a bit puzzling right now

also: surely `std::string` would be preferable to const char* for filename

Reply:
`pipeline::scalars` is a span, so we expect most uses to be constructed implicitly, and it's designed to be cheap to copy.

If I pass by value in general its probably for one of these two reasons:
- The thing is cheap to copy or designed to be constructed implicitly at the call site, e.g. `span`.
- The function is going to make its own copy anyways. In this case, it's better to pass by value, because rvalues won't need to be copied at all this way.